core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -176,14 +173,15 @@ private void completeRouter() {
   }
 
   private CoreEvent eventWithCurrentContextDeleted(CoreEvent event) {
-    Map<String, ForeachContext> foreachContextContainer = foreachContextResolver.getCurrentContextFromEvent(event);
-    foreachContextContainer.remove(event.getContext().getId());
-    return foreachContextResolver.eventWithContext(event, foreachContextContainer);
+    removeContext(event);
+    return quickCopy(event, new HashMap<>(0));
   }
 
   private CoreEvent prepareEvent(CoreEvent event, String expression) {
-    CoreEvent responseEvent =
-        builder(event).addVariable(owner.getRootMessageVariableName(), event.getMessage(), MULE_MESSAGE).build();
+    CoreEvent responseEvent = InternalEvent.builder(event)
+        .addVariable(owner.getRootMessageVariableName(), event.getMessage(), MULE_MESSAGE)
+        .internalParameters(new HashMap<>())
```

Comment:
what is the purpose of cleaning internalParameters here?

Reply:
That was a mistake, internalParameters should not be cleaned